### ISSUE TITLE
ssh directly returns to blink-console without showing error

### ISSUE THREAD

**Issue by mfridge** [blinksh/blink#791]
Dear blink team,

I configured an ssh host and added a private key in config. When connecting to the respective host, nothing happens and I am directly returned to the blink console.

Below you can find the output of "ssh -vvvvvvv". Why can't I connect to the server? Am I missing any configuration? I touched ".ssh/known_hosts" as there is a message about this file to be missing, but I guess that is not the issue here.

Thanks for you help,
Michael

```
blink> ssh -vvvvvvvvv <hostname>
ssh_connect: libssh 0.9.0 (c) 2003-2019 Aris Adamantiadis, Andreas Schneider and libssh contributors. Distributed under the LGPL, please refer to COPYING file for information about your rights, using threading threads_pthread
ssh_connect: Socket connecting, now waiting for the callbacks to work
-[IO stream:handleEvent:]: Received POLLOUT in connecting state
socket_callback_connected: Socket connection callback: 1 (0)
ssh_connect: current state : 2
callback_receive_banner: Received banner: SSH-2.0-OpenSSH_7.1
ssh_client_connection_callback: SSH server banner: SSH-2.0-OpenSSH_7.1
ssh_analyze_banner: Analyzing banner: SSH-2.0-OpenSSH_7.1
ssh_analyze_banner: We are talking to an OpenSSH client version: 7.1 (70100)
ssh_known_hosts_read_entries: Failed to open the known_hosts file '/etc/ssh/ssh_known_hosts': No such file or directory
packet_send2: packet: wrote [type=20, len=1260, padding_size=8, comp=1251, payload=1251]
ssh_send_kex: SSH_MSG_KEXINIT sent
ssh_packet_socket_callback: packet: read type 20 [len=1036,padding=10,comp=1025,payload=1025]
ssh_packet_process: Dispatching handler for packet type 20
ssh_kex_select_methods: Negotiated curve25519-sha256@libssh.org,ssh-ed25519-cert-v01@openssh.com,aes256-gcm@openssh.com,aes256-gcm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-256-etm@openssh.com,zlib@openssh.com,zlib@openssh.com,,
packet_send2: packet: wrote [type=30, len=44, padding_size=6, comp=37, payload=37]
ssh_connect: current state : 6
ssh_packet_socket_callback: packet: read type 31 [len=1412,padding=8,comp=1403,payload=1403]
ssh_packet_process: Dispatching handler for packet type 31
packet_send2: packet: wrote [type=21, len=12, padding_size=10, comp=1, payload=1]
crypt_set_algorithms2: Set output algorithm to aes256-gcm@openssh.com
crypt_set_algorithms2: Set HMAC output algorithm to aead-gcm
crypt_set_algorithms2: Set input algorithm to aes256-gcm@openssh.com
crypt_set_algorithms2: Set HMAC input algorithm to aead-gcm
ssh_init_rekey_state: Set rekey after 4294967296 blocks
ssh_init_rekey_state: Set rekey after 4294967296 blocks
ssh_packet_client_curve25519_reply: SSH_MSG_NEWKEYS sent
ssh_packet_socket_callback: Processing 16 bytes left in socket buffer
ssh_packet_socket_callback: packet: read type 21 [len=12,padding=10,comp=1,payload=1]
ssh_packet_process: Dispatching handler for packet type 21
ssh_packet_newkeys: Received SSH_MSG_NEWKEYS
ssh_packet_newkeys: Signature verified and valid
ssh_connect: current state : 7
blink: setting socket keepalive: 1
blink>
```

**Comment by Schr3da** [comment]
Hey, 

Can you please make sure that you have the Key properly picked in your Host settings?

In general when adding a new Host you should fill out the form as follows (Please pay attention that I make only a pseudo config)

Host: Your_Host_Name
HostName: IP_Address
Port: Port_of_ssh (ssh default 22)
Password: YOUR_PASSWORD
Key: Pick_your_key_from_list (otherwise add a new rsa key if you need any

Thats is usually
Kind regards
Schr3da

**Comment by mfridge** [comment]
Hi,

thanks for your reply! I think I got all this correctly. I also tried another server with another private key, which results in a different behavior. I get asked for a password - probably the private key password (see logs below)?! - an then blink returns to the blink-console. Similar to my first try, this is a password-less key. Could there be an issue with private keys without passwords?

I also checked the imported key using the "copy private key" button in the "key info" page - they are correctly imported.

Best regards,
Michael

```
blink> ssh -vvvvvvv <hostname>
ssh_connect: libssh 0.9.0 (c) 2003-2019 Aris Adamantiadis, Andreas Schneider and libssh contributors. Distributed under the LGPL, please refer to COPYING file for information about your rights, using threading threads_pthread
ssh_connect: Socket connecting, now waiting for the callbacks to work
-[IO stream:handleEvent:]: Received POLLOUT in connecting state
socket_callback_connected: Socket connection callback: 1 (0)
ssh_connect: current state : 2
callback_receive_banner: Received banner: SSH-2.0-OpenSSH_7.9
ssh_client_connection_callback: SSH server banner: SSH-2.0-OpenSSH_7.9
ssh_analyze_banner: Analyzing banner: SSH-2.0-OpenSSH_7.9
ssh_analyze_banner: We are talking to an OpenSSH client version: 7.9 (70900)
ssh_known_hosts_read_entries: Failed to open the known_hosts file '/etc/ssh/ssh_known_hosts': No such file or directory
ssh_client_select_hostkeys: Changing host key method to "ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa,ssh-dss"
packet_send2: packet: wrote [type=20, len=980, padding_size=10, comp=969, payload=969]
ssh_send_kex: SSH_MSG_KEXINIT sent
ssh_packet_socket_callback: packet: read type 20 [len=1076,padding=6,comp=1069,payload=1069]
ssh_packet_process: Dispatching handler for packet type 20
ssh_kex_select_methods: Negotiated curve25519-sha256,ssh-ed25519,aes256-gcm@openssh.com,aes256-gcm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-256-etm@openssh.com,zlib@openssh.com,zlib@openssh.com,,
packet_send2: packet: wrote [type=30, len=44, padding_size=6, comp=37, payload=37]
ssh_connect: current state : 6
ssh_packet_socket_callback: packet: read type 31 [len=188,padding=8,comp=179,payload=179]
ssh_packet_process: Dispatching handler for packet type 31
packet_send2: packet: wrote [type=21, len=12, padding_size=10, comp=1, payload=1]
crypt_set_algorithms2: Set output algorithm to aes256-gcm@openssh.com
crypt_set_algorithms2: Set HMAC output algorithm to aead-gcm
crypt_set_algorithms2: Set input algorithm to aes256-gcm@openssh.com
crypt_set_algorithms2: Set HMAC input algorithm to aead-gcm
ssh_init_rekey_state: Set rekey after 4294967296 blocks
ssh_init_rekey_state: Set rekey after 4294967296 blocks
ssh_packet_client_curve25519_reply: SSH_MSG_NEWKEYS sent
ssh_packet_socket_callback: Processing 196 bytes left in socket buffer
ssh_packet_socket_callback: packet: read type 21 [len=12,padding=10,comp=1,payload=1]
ssh_packet_process: Dispatching handler for packet type 21
ssh_packet_newkeys: Received SSH_MSG_NEWKEYS
ssh_packet_newkeys: Signature verified and valid
ssh_packet_socket_callback: Processing 180 bytes left in socket buffer
ssh_packet_socket_callback: packet: read type 7 [len=160,padding=18,comp=141,payload=141]
ssh_packet_process: Dispatching handler for packet type 7
ssh_packet_ext_info: Received SSH_MSG_EXT_INFO
ssh_packet_ext_info: Follows 1 extensions
ssh_packet_ext_info: Extension: server-sig-algs=<ssh-ed25519,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521>
ssh_connect: current state : 7
blink: setting socket keepalive: 1
packet_send2: packet: wrote [type=5, len=32, padding_size=14, comp=17, payload=17]
ssh_service_request: Sent SSH_MSG_SERVICE_REQUEST (service ssh-userauth)
ssh_packet_socket_callback: packet: read type 6 [len=32,padding=14,comp=17,payload=17]
ssh_packet_process: Dispatching handler for packet type 6 s
ssh_service_request: Sent SSH_MSG_SERVICE_REQUEST (service ssh-userauth)
ssh_packet_socket_callback: packet: read type 6 [len=32,padding=14,comp=17,payload=17]
ssh_packet_process: Dispatching handler for packet type 6
ssh_packet_service_accept: Received SSH_MSG_SERVICE_ACCEPT
packet_send2: packet: wrote [type=50, len=48, padding_size=12, comp=35, payload=35]
ssh_packet_socket_callback: packet: read type 51 [len=48,padding=11,comp=36,payload=36]
ssh_packet_process: Dispatching handler for packet type 51
ssh_packet_userauth_failure: Access denied for 'none'. Authentication that can continue: publickey,keyboard-interactive
ssh_packet_userauth_failure: Access denied for 'none'. Authentication that can continue: publickey,keyboard-interactive
blink: using auth methods attempt: 1 of 5
blink: import key Scaleway
ssh_pki_import_privkey_base64: Trying to decode privkey passphrase=false
ssh_key_algorithm_allowed: Checking rsa-sha2-512 with list <ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ssh-dss-cert-v01@openssh.com,ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa,ssh-dss>
ssh_key_algorithm_allowed: Checking rsa-sha2-512 with list <ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ssh-dss-cert-v01@openssh.com,ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa,ssh-dss>
packet_send2: packet: wrote [type=50, len=608, padding_size=11, comp=596, payload=596]
ssh_packet_socket_callback: packet: read type 51 [len=48,padding=11,comp=36,payload=36]
ssh_packet_process: Dispatching handler for packet type 51
ssh_packet_userauth_failure: Access denied for 'publickey'. Authentication that can continue: publickey,keyboard-interactive
ssh_packet_userauth_failure: Access denied for 'publickey'. Authentication that can continue: publickey,keyboard-interactive
ssh_userauth_kbdint_init: Sending keyboard-interactive init request
packet_send2: packet: wrote [type=50, len=64, padding_size=4, comp=59, payload=59]
ssh_packet_socket_callback: packet: read type 60 [len=48,padding=15,comp=32,payload=32]
ssh_packet_process: Dispatching handler for packet type 60
ssh_packet_userauth_info_request: 1 keyboard-interactive prompts
Password: 
ssh_userauth_kbdint_send: Sending keyboard-interactive response packet
packet_send2: packet: wrote [type=61, len=16, padding_size=6, comp=9, payload=9]
ssh_socket_exception_callback: Socket exception callback: 2 (-1)
ssh_socket_exception_callback: Socket error: Unknown error: -1
blink> 
```

**Comment by yury** [comment]
@mfridge,

It seems that key is wrong. Blink try to auth with public key first (Scaleway), fails and then try to authenticate with keyboard-interactive (password for host, not key)

**Comment by mfridge** [comment]
Ok, I guess I'm too stupid to correctly import the key ;) Is there any best practice how to import the key from the clipboard? Is there maybe another way to import keys instead of going through the clipboard?

**Comment by yury** [comment]
As i understand, you have imported key correctly but server doesn’t accept it.

**Comment by mfridge** [comment]
I import the private key via clipboard. Then, I want to verify that it was imported correctly:

1. I copy the public key to verify it against the original public key --> they are identical
1. I copy the private key to verify it against the original private key --> they are different

This is strange. Is the private key converted to another format in blink?

**Comment by yury** [comment]
yes, we store private key in different format.

**Comment by mfridge** [comment]
Thanks again for you kind support!

I'm afraid that I am somehow at a dead end :-/ The private key is imported correctly - this is confirmed. But I can't authenticate on the server with it. However, I use the same key to connect from my laptop to the same server.

Any idea or any chance to log more details or debug somehow? I use the app from the Appstore and unfortunately I can't compile the open source version myself without larger efforts.

**Comment by yury** [comment]
what type of the key you are using?

**Comment by mfridge** [comment]
RSA with key size 2048 and 4096.

**Comment by yury** [comment]
Can you check that pubkey is in ~/.ssh/authorized_keys file?
Also you can try to copy it with `ssh-copy-id`

**Comment by mfridge** [comment]
Yes, the public key on the server is correctly listed in `authorized_keys`. Copying it again with ssh-copy-id results in
`/usr/bin/ssh-copy-id: WARNING: All keys were skipped because they already exist on the remote system.`
So, the pubkey is correctly on the server. I use the keys (both keys that I tried in blink) to log in to the servers regularly from my PC. So I would say that on the server side everything is setup correctly.

**Comment by yury** [comment]
hmm, just in case, can you try `ssh2` command instead of `ssh`?

**Comment by mfridge** [comment]
Ok, got it. I somehow deleted host-entries in config, re-created them, linked them to newly imported and old keys and it works now.
For one host, it still doesn't work, while the same key allows me to login on another server. I can't find any pattern which might lead to the root cause of all this. I guess I somehow screwed up blinks config?!

Thanks for your help! If I find out some more details, I'll let you know. blink is a great app - once you set it up ;)

**Comment by mfridge** [comment]
@yury I got all my servers correctly configured. I recognized three things that might be helpful with respect to UX design:

1. Port number has to be explicitly set in config. I interpreted the screen, such, that when I leave the port number empty, 22 would be used as default value. This seems not to be the case. It might be helpful if all required fields (hostname/IP, port, username) would be empty by default and are handled as required fields in a way, that you cannot save the server configuration without setting these values. This enforces the user to actively set correct values.
1. As addition to the first point, in my case the default username was more confusing than helpful, because I have different user names at different servers. In one of my server configs, I forgot to set the correct user name, which took some time until I recognized the incorrect setting ;)
1. During my hiccups with the private keys, the solution was to first map another private key to the server config, save this setup and afterwards, map the correct key and save the config again. This was the solution to fix one of my server setups. Sorry, that I cannot describe this more clearly. It seemed to me, that this was the activity that resolved config issues while I tried different keys.

Maybe this can be helpful for further development. Thanks again for your support and your patience :)
